flaskbb/utils/requirements.py

```diff
@@ -113,185 +122,217 @@ def _determine_locked(self, request):
 
-        return set([g.id for g in current_forum.groups]) & set([g.id for g in user.groups])
-
-
-class CanAccessTopic(Requirement):
-    def fulfill(self, user, request):
+    def fulfill(self, user):
         if not current_forum:
-            raise FlaskBBError('Could not load topic data')
+            raise FlaskBBError("Could not load forum data")
 
-        return set([g.id for g in current_forum.groups]) & set([g.id for g in user.groups])
+        forum_groups = {g.id for g in current_forum.groups}
+        user_groups = {g.id for g in user.groups}
+        return bool(forum_groups & user_groups)
 
 
 def IsAtleastModeratorInForum(forum_id=None, forum=None):
-    return Or(IsAtleastSuperModerator, IsModeratorInForum(forum_id=forum_id,
-                                                          forum=forum))
+    return Or(
+        IsAtleastSuperModerator,
+        IsModeratorInForum(forum_id=forum_id, forum=forum),
+    )
 
 
-IsMod = And(IsAuthed(), Has('mod'))
-IsSuperMod = And(IsAuthed(), Has('super_mod'))
-IsAdmin = And(IsAuthed(), Has('admin'))
+IsMod = And(IsAuthed(), Has("mod"))
+IsSuperMod = And(IsAuthed(), Has("super_mod"))
+IsAdmin = And(IsAuthed(), Has("admin"))
 
 IsAtleastModerator = Or(IsAdmin, IsSuperMod, IsMod)
 
 IsAtleastSuperModerator = Or(IsAdmin, IsSuperMod)
 
-CanBanUser = Or(IsAtleastSuperModerator, Has('mod_banuser'))
+CanBanUser = Or(IsAtleastSuperModerator, Has("mod_banuser"))
 
-CanEditUser = Or(IsAtleastSuperModerator, Has('mod_edituser'))
+CanEditUser = Or(IsAtleastSuperModerator, Has("mod_edituser"))
 
-CanEditPost = Or(IsAtleastSuperModerator,
-                 And(IsModeratorInForum(), Has('editpost')),
-                 And(IsSameUser(), Has('editpost'), TopicNotLocked()))
+CanEditPost = Or(
+    IsAtleastSuperModerator,
+    And(IsModeratorInForum(), Has("editpost")),
+    And(CanAccessForum(), IsSameUser(), Has("editpost"), TopicNotLocked()),
+)
 
 CanDeletePost = CanEditPost
 
-CanPostReply = Or(And(Has('postreply'), TopicNotLocked()),
-                  IsModeratorInForum(),
-                  IsAtleastSuperModerator)
+CanPostReply = Or(
+    And(Has("postreply"), TopicNotLocked()),
+    IsModeratorInForum(),
+    IsAtleastSuperModerator,
+)
 
-CanPostTopic = Or(And(Has('posttopic'), ForumNotLocked()),
-                  IsAtleastSuperModerator,
-                  IsModeratorInForum())
+CanPostTopic = Or(
+    And(Has("posttopic"), ForumNotLocked()),
+    IsAtleastSuperModerator,
+    IsModeratorInForum(),
+)
 
-CanDeleteTopic = Or(IsAtleastSuperModerator,
-                    And(IsModeratorInForum(), Has('deletetopic')),
-                    And(IsSameUser(), Has('deletetopic'), TopicNotLocked()))
+CanDeleteTopic = Or(
+    IsAtleastSuperModerator,
+    And(IsModeratorInForum(), Has("deletetopic")),
+    And(IsSameUser(), Has("deletetopic"), TopicNotLocked()),
+)
 
 
-# Template Allowances -- gross, I know
+def permission_with_identity(requirement, name=None):
```

Comment:
In 0.6.0 Permission can be constructed outside of an active context which makes it nicer for our use cases. This factory makes for a nice use case with jinja2 filters that act on a user.